src/services/bunker_cases/midterm_slashing_penalty.py

```diff
@@ -0,0 +1,190 @@
+
+        return per_epoch_buckets
+
+    @staticmethod
+    def not_withdrawn_slashed_validators(
+        all_validators: Mapping[str, Validator], ref_epoch: EpochNumber
+    ) -> dict[str, Validator]:
+        """
+        Get all slashed validators, who are not withdrawn yet
+        """
+        slashed_validators: dict[str, Validator] = {}
+
+        for key, v in all_validators.items():
+            if v.validator.slashed and int(v.validator.withdrawable_epoch) > ref_epoch:
+                slashed_validators[key] = v
+
+        return slashed_validators
+
+    @staticmethod
+    def get_per_epoch_lido_midterm_penalties(
+        per_epoch_buckets: dict[EpochNumber, dict[str, Validator]],
+        lido_slashed_validators: dict[str, Validator],
+        total_balance: Gwei,
+    ) -> dict[EpochNumber, dict[str, Gwei]]:
+        """
+        Iterate through per_epoch_buckets and calculate lido midterm penalties for each bucket
+        """
+        per_epoch_lido_midterm_penalties: dict[EpochNumber, dict[str, Gwei]] = defaultdict(dict)
+        for key, v in lido_slashed_validators.items():
+            midterm_penalty_epoch = MidtermSlashingPenalty.get_midterm_slashing_epoch(v)
+            # We should calculate midterm penalties by sum of slashings which bound with midterm penalty epoch
+            bound_slashed_validators = MidtermSlashingPenalty.get_bound_slashed_validators(
+                per_epoch_buckets, midterm_penalty_epoch
+            )
+            slashings = sum(int(v.validator.effective_balance) for v in bound_slashed_validators.values())
+            adjusted_total_slashing_balance = min(
+                slashings * PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX,
+                total_balance
+            )
+            effective_balance = int(v.validator.effective_balance)
+            penalty_numerator = effective_balance // EFFECTIVE_BALANCE_INCREMENT * adjusted_total_slashing_balance
+            penalty = penalty_numerator // total_balance * EFFECTIVE_BALANCE_INCREMENT
+            per_epoch_lido_midterm_penalties[midterm_penalty_epoch][key] = Gwei(penalty)
+        return per_epoch_lido_midterm_penalties
+
+    @staticmethod
+    def get_per_frame_lido_midterm_penalties(
+        per_epoch_lido_midterm_penalties: dict[EpochNumber, dict[str, Gwei]],
+        frame_config: FrameConfig,
+    ) -> list[Gwei]:
+        """
+        Put per epoch buckets into per frame buckets to calculate lido midterm penalties impact in each frame
+        """
+        per_frame_buckets: dict[int, dict[str, Gwei]] = defaultdict(dict)
+        for epoch, validator_penalty in per_epoch_lido_midterm_penalties.items():
+            frame_index = MidtermSlashingPenalty.get_frame_by_epoch(epoch, frame_config)
+            for val_key, penalty in validator_penalty.items():
+                per_frame_buckets[frame_index][val_key] = (
+                    max(penalty, per_frame_buckets[frame_index].get(val_key, Gwei(0)))
+                )
```

Comment:
why it's possible to have more than one validator with the same pubkey here? if i understand right, it's groupped by midterm epoch, and it can be only one for each slashing

Reply:
fixed